src/Renci.SshNet/SftpClient.cs

```diff
@@ -476,6 +542,62 @@ public IEnumerable<SftpFile> ListDirectory(string path, Action<int> listCallback
             return InternalListDirectory(path, listCallback);
         }
 
+#if FEATURE_TAP
+
+        /// <summary>
+        /// Asynchronously retrieves list of files in remote directory.
+        /// </summary>
+        /// <param name="path">The path.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>
+        /// A <see cref="Task{IEnumerable}"/> that represents the asynchronous list operation.
+        /// The task result contains an enumerable collection of <see cref="SftpFile"/> for the files in the directory specified by <paramref name="path" />.
+        /// </returns>
+        /// <exception cref="ArgumentNullException"><paramref name="path" /> is <b>null</b>.</exception>
+        /// <exception cref="SshConnectionException">Client is not connected.</exception>
+        /// <exception cref="SftpPermissionDeniedException">Permission to list the contents of the directory was denied by the remote host. <para>-or-</para> A SSH command was denied by the server.</exception>
+        /// <exception cref="SshException">A SSH error where <see cref="Exception.Message" /> is the message from the remote host.</exception>
+        /// <exception cref="ObjectDisposedException">The method was called after the client was disposed.</exception>
+        public async Task<IEnumerable<SftpFile>> ListDirectoryAsync(string path, CancellationToken cancellationToken)
```

Comment:
Consider returning `ICollection` rather than `IEnumerable` as per the [Microsoft Guidelines for Collections](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections).

> DO use Collection<T> or a subclass of Collection<T> for properties or return values representing read/write collections.

Reply:
This should be **IAsyncEnumerable\<SftpFile>**. It's only supported on .NET 4.6.1 or higher.
For .NET 4.6.1 and .NET Standard 2.0, you need to add a reference to **Microsoft.Bcl.AsyncInterfaces**.

Reply:
Then we need to remove FEATURE_TAP from netstandard1.3. Also, we take two "external" dependencies (AsyncInterfaces and Task.Extensions).
What if we leave this one as it is and add ```IAsyncEnumerable<SftpFile> EnumerateDirectoryAsync(...)``` for netstandard2.1+ ?

Reply:
I won't let this block this merge request, but I still strongly consider changing this method to return **IAsyncEnumerable\<SftpFile>**. I also don't like bringing in additional dependencies, but the alternative is not better.